library/tiqr/Tiqr/UserStorage/GenericStore.php

```diff
@@ -25,104 +25,115 @@
  * please supply your own implementation that hosts the data in your user database OR
  * in a secure (e.g. hardware encrypted) storage.
  * @author ivo
+ *
+ * Note: This implementation does not employ locking or transactions, and is not safe when the same user is updated
+ *       concurrently
  */
 abstract class Tiqr_UserStorage_GenericStore extends Tiqr_UserStorage_Abstract
 {
 
-    abstract protected function _loadUser($userId, $failOnNotFound = TRUE);
-  
-    abstract protected function _saveUser($userId, $data);
+    /** Return data of user $userId
+     * @param string $userId
+     * @return array
+     * @throws Exception
+     */
+    abstract protected function _loadUser(string $userId): array;
+
+    /** Write data of $userId
+     * @param string $userId
+     * @param array $data
+     * @throws Exception
+     */
+    abstract protected function _saveUser(string $userId, array $data): void;
+
+    /**
+     * @param string $userId
+     * @return bool trues when user exists, false otherwise
+     * @throws Exception
+     */
+    abstract protected function _userExists(string $userId): bool;
 
 
     /**
-     * (non-PHPdoc)
-     * @see simplesamlphp/modules/authTiqr/lib/User/sspmod_authTiqr_User_Interface::createUser()
+     * @see Tiqr_UserStorage_Interface::createUser()
      */
-    public function createUser($userId, $displayName) 
+    public function createUser(string $userId, string $displayName) : void
     {
         $user = array("userId"=>$userId,
                       "displayName"=>$displayName);
-        return $this->_saveUser($userId, $user);
+        $this->_saveUser($userId, $user);
     }
 
     /**
-     * (non-PHPdoc)
-     * @see simplesamlphp/modules/authTiqr/lib/User/sspmod_authTiqr_User_Interface::userExists()
+     * @see Tiqr_UserStorage_Interface::userExists()
      */
-    public function userExists($userId)
+    public function userExists(string $userId): bool
     {
-        $user = $this->_loadUser($userId, FALSE);
-        return (is_array($user)||is_object($user)); 
+        return $this->_userExists($userId);
     }
 
     /**
-     * (non-PHPdoc)
-     * @see simplesamlphp/modules/authTiqr/lib/User/sspmod_authTiqr_User_Interface::getDisplayName()
+     * @see Tiqr_UserStorage_Interface::getDisplayName()
      */
-    public function getDisplayName($userId) 
+    public function getDisplayName(string $userId): string
     {
         if ($data = $this->_loadUser($userId)) {
             return $data["displayName"];
         }
-        return NULL;
+        return '';
```

Comment:
NULL vs '' makes the method type safe :+1: 

Would throwing an exception when the display name is not found not somewhat better than returning an empty string? I know this is context dependent.

Reply:
You're correct that this could never happen because createUser sets the displayname. So we could treat it as a backend storage error.
The file based implementation is only suitable for testing in practice so I didn't give it as much attention as e.g. the PDO UserStorage.